### DISCUSSION TITLE
GUI Features

### DISCUSSION THREAD

**Discussion by rover-t** [JamesSG2/Handwriting-Conversion-Application#16] [General]
Hey @JosephHaugen here is some information on what feature would be required for the GUI.

Here is the main function where we get input from the user. I've added comments that should give you and idea of what sorts of inputs we will need and how I'm envisioning the user will interact with the program. If you have any questions or suggestions let me know. This code can be found near the bottom of the dev branch in the main.py program.

    os.chdir("..")
    print("Analyze new samples? [Y/N]:")
    analysis = input() 
    //first we ask if we are analyzing a new sample or using an old one. This can return anything we can use for the if elif
    //statement that follows

    if(analysis=="Y"):
        print("Name: ")
        name = input()
        //we want their name to know where to store the analyzed data
        print("Image sample: ")
        target = input()
        //get the name of the actual image file. 
        //This could be a string like it is now, or we can modify it to take a full file path.

        img = analyze_sample(name, target)
        cv2.imshow("full", img)
        cv2.waitKey(0)
        //it would be nice if we could show these images directly in the GUI, not very important though

    elif(analysis=="N"):
        print("Name: ")
        name = input()
        //once again we want their name to know where to look for the analyzed data
        print("What should be written:")
        phrase = input()
        //this should be the text string that they want to recreate in their handwriting

        img = output_handwriting(name, phrase)
        cv2.imshow("output", img)
        cv2.waitKey(0)
        //once again it would be nice if we could show these images directly in the GUI, not very important though

    else:
        print("Error, did not select [Y/N]")
        //it would be nice if there to log errors and updates besides printing to the command line

    print("Done")

**Comment by JosephHaugen** [comment]
I think I got what you needed, its kinda ugly but it should get the job done for now. I committed the changes to GUI1.py in the GUI branch. If you want to test it you wont need to install any libraries so you can just download the code. Let me know if I messed anything up or there is something else you need to have.

**Reply by rover-t** [reply]
Sounds great! I'll check it out tomorrow and let you know how it goes.

**Comment by BigWub67** [comment]
I came up with an idea for how the whole app will generally look and function. The grey highlight is where that button from the start screen takes you.

![gui idea](https://user-images.githubusercontent.com/80499657/113970037-f4cab980-97ea-11eb-903d-bea60818abd6.png)

**Reply by rover-t** [reply]
That looks great!

**Comment by BigWub67** [comment]
@JosephHaugen Here's is the plan I've come up with to make good progress with the GUI in the next 7 days:

I'll make the screen manager to connect all the screens together and make the two screens for the right button.

You make the four screens for the left button. Have them all be separate scripts. For functionality, just have all the widgets there so I could later make it look better. For the buttons (like back, continue, home), just give them a command that is named appropriately (like toStartScreen() for the home button). I'll add in the part to transition between screens after I have the screen manager made. If you get that done quickly, you could either start making them look good or start on the two screens I'm supposed to make (which I would then focus more on making other four screens look good).

Let me know if you want to go about it another way or something.

**Reply by JosephHaugen** [reply]
That plan sounds great, my only concern is time. I've never really written multiple python scripts that would work together before so I would need some time to learn. I like your UI layout though, I was wondering if it would be simpler to just modify my code, to make it look nice, as I already have most of the functionality we need. You know more about both ideas already so I think you should decide how we proceed, just keep in mind I wouldn't be much help initially and we are on a bit of a time crunch.

**Reply by rover-t** [reply]
You mentioned changing the name of the main.py file. Would backend.py be a good new name?

**Comment by BigWub67** [comment]
@JosephHaugen I made the screen manager so you don't really have to worry about making the separate scripts work together. You can see how to transition to other screens by looking at how I did it with startScreen.py.

For modifying your code, I was already planning on making it look better later. It will also need a lot of modifying to work as a multiscreen app (that also resizes widgets with window). Ideally you would learn to make it look good and resize when first making it though.

Now I'll go into how I'm thinking the profile system will work:

When you create a profile (user types in name and hit continue), it creates an empty folder in the output folder. When finding all profiles to display, it will search for every folder in output and display a button for every folder. Each folder is a profile, so if a folder's called 'trevor' in output then it will display a button with the text 'trevor'. If there are a lot of profiles that it doesn't show them all, it will then show a scroll wheel for the user to scroll down the list of buttons.

For what you should now do with your current code:

1. Change:
from tkinter import *
to: 
import tkinter as tk
and do all the changes necessary for it to function since wildcard importing should be avoided.

2. Add:
from tkinter import ttk
to have more modern looking buttons (and also actually use it with the buttons, look at my code as reference)

3. Make it a class like how I've done with the start screen. 

I'd also like if you fix the spelling mistakes and make the gui significantly more clear about what's needed and what buttons do.

**Comment by JosephHaugen** [comment]
Hey @BigWub67 , I built the upload screen and committed it to the GUI folder. I didn't add the continue and back button functionality as my code for that was very buggy and did not work well with your method. I'm sorry I can't help all that much but I this screen can select multiple files and store them in an array to be used in trevor's code. I left out any styling for the screen as I know you have a plan for what you want everything to look like. let me know if you need me to get anything else done.

**Comment by BigWub67** [comment]
@JamesSG2 and @rover-t

I got a early somewhat working version of the generate handwriting screen. It's using the backend.py from the main branch with a small alteration to not create more than 1 result file.

Right now I don't have the part to remove bad letters, it doesn't wrap, and you can't save it somewhere. 

I don't know which version of backend.py I should be using for it, so if one of you could let me know that would be great.

I feel we should communicate about how to implement the removing letters part, since that seems to be the last big hurdle.

In the meantime, I will work on the other parts of the GUI.

**Reply by BigWub67** [reply]
@JamesSG2 and @rover-t 

I've made almost all of the GUI, here's what's left (or at least what I could think of):

Spellcheck - I had to remove (comment out for now) spell check in backend because it wouldn't work with how file paths are given
Letter Correction - Like I mentioned in previous comment
Multi line images - Like I mentioned in previous comment

**Reply by rover-t** [reply]
The dev branch should have the most updated version of the backend. It will include all the features you mentioned.

**Reply by rover-t** [reply]
Could you elaborate on why the spellcheck isn't working? That part, along with Jame's error level, does quite a bit to improve accuracy.

**Reply by BigWub67** [reply]
The reason why the spellcheck doesn't work is because it passes the entire path to the image, not just the name.

**Reply by rover-t** [reply]
Got it, I think I solved the issue. There should be a pull request with a fix if you want to check that.
